integration/querydsl/blaze-persistence-querydsl-testsuite/src/test/java/com/blazebit/persistence/querydsl/BasicQueryTest.java

```diff
@@ -0,0 +1,744 @@
+        });
+    }
+
+    // NOTE: Entity joins are only supported on Hibernate 5.1+
+    // NOTE: Hibernate 5.1 renders t.id = tSub.id rather than t = tSub
+    @Test
+    @Category({ NoDatanucleus.class, NoEclipselink.class, NoOpenJPA.class, NoHibernate42.class, NoHibernate43.class, NoHibernate50.class, NoHibernate51.class})
+    public void testMultipleInlineEntityWithLimitJoin() {
+        doInJPA(entityManager -> {
+            QRecursiveEntity t = new QRecursiveEntity("t");
+            QRecursiveEntity subT = new QRecursiveEntity("subT");
+            QRecursiveEntity subT2 = new QRecursiveEntity("subT2");
+
+            List<RecursiveEntity> fetch = new BlazeJPAQuery<RecursiveEntity>(entityManager, cbf)
+                    .select(t)
+                    .from(new BlazeJPAQuery<RecursiveEntity>().select(t).from(t)
+                            .leftJoin(selectFrom(subT).where(subT.parent.name.eq("root1")).orderBy(subT.name.asc()).limit(1), subT)
+                            .on(t.eq(subT))
+                            .where(t.parent.name.eq("root1"))
+                            .orderBy(t.name.asc())
+                            .limit(1L), t)
+                    .leftJoin(selectFrom(subT2)
+                            .where(subT2.parent.name.eq("root1"))
+                            .orderBy(subT2.name.asc())
+                            .limit(1), subT2)
+                    .on(t.eq(subT2))
+                    .fetch();
+
+            assertNotNull(fetch);
+
+        });
+    }
+
+    @Test
+    @Category({ NoDatanucleus.class, NoEclipselink.class, NoOracle.class, NoOpenJPA.class, NoHibernate42.class, NoHibernate43.class, NoHibernate50.class, NoHibernate51.class, NoH2.class, NoMySQLOld.class })
+    public void testMultipleInlineEntityWithLimitJoinLateral() {
+        Assume.assumeFalse(dbmsDialect.getLateralStyle() == LateralStyle.NONE);
+
+        doInJPA(entityManager -> {
+            QRecursiveEntity t = new QRecursiveEntity("t");
+            QRecursiveEntity subT = new QRecursiveEntity("subT");
+            QRecursiveEntity subT2 = new QRecursiveEntity("subT2");
+            QRecursiveEntity subT3 = new QRecursiveEntity("subT3");
+
+            new BlazeJPAQuery<RecursiveEntity>(entityManager, cbf)
+                    .select(t)
+                    .from(new BlazeJPAQuery<RecursiveEntity>().select(t).from(t)
+                            .leftJoin(selectFrom(subT).where(subT.parent.name.eq("root1")).orderBy(subT.name.asc()).limit(1), subT)
+                            .on(t.eq(subT))
+                            .where(t.parent.name.eq("root1"))
+                            .orderBy(t.name.asc())
+                            .limit(1L), t)
+                    .leftJoin(selectFrom(subT2)
+                        .where(subT2.parent.name.eq("root1"))
+                        .orderBy(subT2.name.asc())
+                        .limit(1), subT3)
+                        .on(t.eq(subT3))
+                        .lateral()
```

Comment:
Just thinking out loud, wouldn't it be better to have methods like `leftJoinLateral` etc. since lateral can only be left or inner?

Reply:
I've used `fetchJoin()` and `fetchAll()` as comparison material here. There methods suffer from the same issue (as these are only applicable to association joins and not entity joins). Its not the most ideal API but I also do value some consistency here. Not sure though how this way of dealing with `JoinFlags` was ever well thought through. There aren't so many other types of JoinFlags around in QueryDsl to compare with.

Reply:
I understand that you want to stay as close to the QueryDSL model as possible.